### PR TITLE
Feature/spat msg cpp

### PR DESCRIPTION
**Pull request by adev4a**
# PR Details
This PR adds the message conversion for SPaT messages. 
## Description
It should be noted that this implementation only successfully tested Decoding of SPaT messages.
The encoding implementation and unit test will be completed under https://github.com/usdot-fhwa-stol/carma-platform/issues/1382  

## How Has This Been Tested?

Unit Tested

## Types of changes

- [ ] Defect fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [ ] I have added any new packages to the sonar-scanner.properties file
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
[CARMA Contributing Guide](https://github.com/usdot-fhwa-stol/carma-platform/blob/develop/Contributing.md) 
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: carma-messenger-core/cpp_message/src/SPAT_Message.cpp [hunk 1]
```diff
@@ -0,0 +1,702 @@
...
+
+                    //State Time Speed Movement Event List
+                    j2735_msgs::MovementEventList movement_event_list;
+                    for (int k = 0; k < state->states.list.array[j]->state_time_speed.list.count; k++)
+                    {
+                        if(!state->states.list.array[j]->state_time_speed.list.array[k]){
+                            continue;
+                        }
+                        j2735_msgs::MovementEvent movement_event;
+                        //movement_event.event_state = state->states.list.array[j]->state_time_speed.list.array[j]->eventState;
```
Comment (MishkaMN) [inline carma-messenger-core/cpp_message/src/SPAT_Message.cpp:219]: commented code

File: carma-messenger-core/cpp_message/src/SPAT_Message.cpp [hunk 2]
```diff
@@ -0,0 +1,702 @@
...
+                        //Decode movement event
+                        //1. MovementPhaseState
+                        if(state->states.list.array[j]->state_time_speed.list.array[k]->eventState){
+                            movement_event.event_state.movement_phase_state = state->states.list.array[j]->state_time_speed.list.array[k]->eventState;
```
Comment (MishkaMN) [inline carma-messenger-core/cpp_message/src/SPAT_Message.cpp:223]: I think eventState is uin8_t, and DARK=1 for example is `01000000` which is 64 in int not 1. 
So I think we want similar decoding like you did in `//Decode status`

File: carma-messenger-core/cpp_message/src/SPAT_Message.cpp [hunk 3]
```diff
@@ -0,0 +1,702 @@
...
+                                j2735_msgs::AdvisorySpeed advisory_speed;
+
+                                if (state->states.list.array[j]->state_time_speed.list.array[k]->speeds->list.array[l]->type)
+                                {
+                                    advisory_speed.type.advisory_speed_type = state->states.list.array[j]->state_time_speed.list.array[k]->speeds->list.array[l]->type;
+                                }
+                                else
+                                {
+                                    advisory_speed.type.advisory_speed_type = ADVISORY_SPEED_TYPE_NONE_;
+                                }
```
Comment (MishkaMN) [inline carma-messenger-core/cpp_message/src/SPAT_Message.cpp:300]: advisory speed type is not optional it seems.
https://github.com/usdot-fhwa-stol/carma-msgs/blob/develop/j2735_msgs/msg/AdvisorySpeed.msg
and advisory_speed_type is also in bits, so ==1 check is needed I think

**Review by MishkaMN — CHANGES_REQUESTED** [review]

File: carma-messenger-core/cpp_message/src/SPAT_Message.cpp [hunk 4]
```diff
@@ -0,0 +1,700 @@
...
+                                timing.next_time = *next_time;
+                            }
+                            movement_event.timing = timing;
+                        }
+                        //3. Advisory Speed List
+                        movement_event.speeds_exists = false;
+                        if (state->states.list.array[j]->state_time_speed.list.array[k]->speeds)
+                        {
+                            movement_event.speeds_exists = true;
+                            j2735_msgs::AdvisorySpeedList advisory_speed_list;
```
Comment (MishkaMN) [inline carma-messenger-core/cpp_message/src/SPAT_Message.cpp:288]: looks like we are not using this

**Review by MishkaMN — CHANGES_REQUESTED** [review]

File: carma-messenger-core/cpp_message/src/SPAT_Message.cpp (hunk 2 above)
Comment (adev4a) [reply carma-messenger-core/cpp_message/src/SPAT_Message.cpp:223]: For this eventState is a uint8_t not a bit array like status was. Checking the definition eventState is a MovementPhaseState which is defined as an enum ranging from 0-9
https://github.com/usdot-fhwa-stol/carma-messenger/blob/9d3a864dab2bd6bf0762ba954d1c3ed744fbc83c/carma-messenger-core/cpp_message/include/generated/MovementEvent.h#L31
https://github.com/usdot-fhwa-stol/carma-messenger/blob/develop/carma-messenger-core/cpp_message/include/generated/MovementPhaseState.h#L22

File: carma-messenger-core/cpp_message/src/SPAT_Message.cpp (hunk 3 above)
Comment (adev4a) [reply carma-messenger-core/cpp_message/src/SPAT_Message.cpp:300]: By not optional do you mean there should be no if condition checking if a value exists?

Advisory speed type is also a single bit value from an enum https://github.com/usdot-fhwa-stol/carma-messenger/blob/develop/carma-messenger-core/cpp_message/include/generated/AdvisorySpeedType.h

Comment (MishkaMN) [reply carma-messenger-core/cpp_message/src/SPAT_Message.cpp]: yeah I don't think we need to check if value exists

File: carma-messenger-core/cpp_message/src/SPAT_Message.cpp (hunk 2 above)
Comment (MishkaMN) [reply carma-messenger-core/cpp_message/src/SPAT_Message.cpp:225]: ah I see that makes sense

**Review by MishkaMN — APPROVED** [review]

**Review by MishkaMN — APPROVED** [review]

**Pull request merged**